Adds Node-compatible protocolForURL

tomdale opened the pull request:
The current implementation of the DOM helper’s `protocolForURL` relies
on the implicit parsing and normalization that the browser’s DOM does
when setting the A element’s `href` property.

However, the DOM helper is often used with subset DOM implementations
like simple-dom. In those cases, we cannot rely on this normalization.

This commit detects when the DOM Helper is running in Node and uses
the built-in URL package to parse the protocol instead.

/cc @wycats @chancancode @rwjblue @stefanpenner

tomdale commented:
It looks like CJS commits were disabled in https://github.com/tildeio/htmlbars/commit/0c51a428dd0aa8358c075d73d82fd90df9d6536f but no rationale for why was provided in that commit. Presumably it was for a good reason, but it means I couldn't write a test for this case.

I think it's OK because this is implicitly tested in Ember anyway and this implementation of the DOM Helper is EOL once Glimmer2 lands.

stefanpenner commented on packages/dom-helper/lib/main.js:
```diff
@@ -567,13 +569,26 @@ prototype.parseHTML = function(html, contextualElement) {
-  }
+// In Node.js environments, we can't rely on the DOM correctly parsing URLs.
+// Instead, we'll fall back to Node's built-in URL parsing.
+if (typeof process === 'object' && process + '' === '[object process]') {
```
does this work in node-webkit?

tomdale replied:
Unknown, but looks like this might be a better test: https://github.com/jnath/nwjs-video-player/blob/159c0d487159846664782979d46bbd61857fb429/src/app/js/nw.js#L1

chancancode replied:
cc @wycats :eyes:

krisselden replied:
`process` is likely not something you want to expose on the sandbox context just for an is node check.

stefanpenner commented:
> It looks like CJS commits were disabled in 0c51a42 but no rationale for why was provided in that commit. Presumably it was for a good reason, but it means I couldn't write a test for this case.

They where breaking the build, and I did not have time to fix them.

chancancode commented:
seems good! :+1:

tomdale commented:
@stefanpenner 
> Maybe someone else can

Basically EOL at this point, not sure it's worth anyone putting effort in here that could be put into glimmer.

stefanpenner commented:
> Basically EOL at this point, not sure it's worth anyone putting effort in here that could be put into glimmer.

@tomdale c

tomdale commented:
@stefanpenner @krisselden @chancancode Replaced detecting the environment with a feature detection. How does this approach look?

stefanpenner commented:
@tomdale I would feel better doing the test before the prototype member is set, rather then replacing the SharedFunction during invocation. Is this possible, or is it lazy fora specific reason?

stefanpenner commented on packages/dom-helper/lib/main.js:
```diff
@@ -565,15 +567,39 @@ prototype.parseHTML = function(html, contextualElement) {
+    // Otherwise, we need to fall back to our own URL parsing.
+    // Global `require` is shadowed by Ember's loader so we have to use the fully
+    // qualified `module.require`.
+    URL = module.require('url');
```
what if there is no require? Might be nice to fail fast then (might be aligned with my comment bellow about doing this before setting up the prototype.

stefanpenner commented on packages/dom-helper/lib/main.js:
```diff
@@ -565,15 +567,39 @@ prototype.parseHTML = function(html, contextualElement) {
+
+function browserProtocolForURL(url) {
   if (!parsingNode) {
     parsingNode = this.document.createElement('a');
```
we may as well just create this node in the outer context, rather then a branch hit once, its just extra weight on the inlining budget.

tomdale replied:
I don't think you can create the node in the outer context because you don't yet have the `document`, which gets passed in by the consumer.

tomdale replied:
Basically you aren't guaranteed to have the `document` until instantiation time.

stefanpenner replied:
then it can be done in the constructor, and rather then overriding the method on the prototype, it could defer to a delegate which changes. This prevents some sadness we have been seeing with jits, and swapping out methods at runtime.


```js

function C() {
   this._prococalForURL = ...;
}

C.prototype.prococalForURL = function(url) {
  this._prococalForURL(url)
}
```

tomdale replied:
What is the benefit of having a prototype method that delegates to an instance method that's been installed at construction time? Wouldn't it just be better to install the method on the instance?

```js
function C() {
  if (isNode) {
    this.protocolForURL = nodeProtocolForURL;
  } else {
    this.protocolForURL = browserProtocolForURL;
  }
}

function nodeProtocolForURL() { ... };
function browserProtocolForURL() { ... };
```

tomdale commented:
@chancancode @stefanpenner I updated the implementation to install the methods on the instance, and also added support for raw HTML sections. That last change let's us remove all of the monkeypatches from Ember/FastBoot.

tomdale commented:
Gonna merge this, @stefanpenner let me know if you've got any issues and I will fix.
